### PR TITLE
Transaction impls use Eclipse primitive collections for in-memory transaction tracking

### PR DESCRIPTION
**Pull request by schlosna**
## General
**Before this PR**:

Was looking through some JFR for a service heavily using AtlasDB queues and noticed a lot of `Long.valueOf` boxing in `com.palantir.atlasdb.transaction.impl.SnapshotTransaction.collectCellsToPostFilter(TableReference, Map, Collection, Function, AsyncKeyValueService, Set, Map):1641` to get commit timestamp for each result value:

https://github.com/palantir/atlasdb/blob/33865f731778e813c41aa03b24eae93de88fdc59/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java#L1641

image

**After this PR**:

==COMMIT_MSG==
Transaction impls use Eclipse primitive collections for in-memory transaction tracking
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:
Needs benchmark & performance testing. Eclipse collections isn't heavily used in AtlasDB yet, there may be other places we'd want to consider using Eclipse primitive collections for more memory efficient collections & data structures.

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 1]
```diff
@@ -2330,30 +2342,36 @@ protected Map<Cell, Long> detectWriteAlreadyCommittedInternal(
...
+                    throw new SafeIllegalArgumentException(
+                            "Cannot rollback already committed transaction",
+                            SafeArg.of("startTs", startTs),
+                            SafeArg.of("commitTs", commitTs));
                 }
-            } else {
-                Preconditions.checkArgument(commitTimestamps.get(startTs) == TransactionConstants.FAILED_COMMIT_TS);
             }
-        }
+            log.warn("Rolling back transaction: {}", SafeArg.of("startTs", startTs));
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:2358]: I think we only want to do this if `startTs` is not in `commitTimestamps` - so this and the rollbackOther needs to be in an else block; or the end of the if-block above should have a `return true;`?

**Review by jeremyk-91 — COMMENTED** [review]
The translation looks mostly right, one question though

**Comment by schlosna** [comment]
Thanks for the review @jeremyk-91 ! Updated to address your comment and fix tests

**Comment by schlosna** [comment]
Saw some significant `Long` boxing in JFRs on another AtlasDB 🎿 ⛰️ services that was also sweeping heavily (see https://github.com/palantir/atlasdb/pull/6676)

https://github.com/palantir/atlasdb/blob/4e050a9cb4f8f2b45b8b700df082f442d2fec694/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/CommitTimestampLoader.java#L134-L161

https://github.com/palantir/atlasdb/blob/4e050a9cb4f8f2b45b8b700df082f442d2fec694/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java#L260-L287

**Review by jeremyk-91 — APPROVED** [review]
👍  Looks great. Thanks!

**Pull request merged**